### PR TITLE
[java] Properly resolve array types

### PR DESCRIPTION
**Pull request by jsotuyod**
- Honor dimensions of arrays
 - Resolve types for allocations as well as declarations of arrays

### CODE REVIEW

File: pmd-java/src/main/java/net/sourceforge/pmd/lang/java/typeresolution/ClassTypeResolver.java [hunk 1]
```diff
@@ -1137,38 +1135,12 @@ public Object visit(ASTLiteral node, Object data) {
...
-            // missing some TypeNode candidates we can add to the AST and have
-            // to deal
-            // with here (e.g. FormalParameter)? Plus some existing usages may
-            // be
-            // incorrect.
+        final ASTArrayDimsAndInits dims = node.getFirstChildOfType(ASTArrayDimsAndInits.class);
+        if (dims != null) {
+            final Class<?> arrayType = ((TypeNode) node.jjtGetChild(0)).getType();
+            if (arrayType != null) {
+                node.setType(Array.newInstance(arrayType, (int[]) Array.newInstance(int.class, dims.getArrayDepth())).getClass());
```
Comment (adangel) [inline pmd-java/src/main/java/net/sourceforge/pmd/lang/java/typeresolution/ClassTypeResolver.java]: Nice, didn't know Array.newInstance before :)

**Comment by adangel** [comment]
@jsotuyod 
I'll mark ASTArrayDimsAndInits and ASTVariableDeclaratorId as [Dimensionable](https://github.com/pmd/pmd/blob/master/pmd-java/src/main/java/net/sourceforge/pmd/lang/java/ast/Dimensionable.java) when I merge it. We use it already on some nodes.

File: pmd-java/src/main/java/net/sourceforge/pmd/lang/java/typeresolution/ClassTypeResolver.java (hunk 1 above)
Comment (adangel) [reply pmd-java/src/main/java/net/sourceforge/pmd/lang/java/typeresolution/ClassTypeResolver.java]: Oh, I used it before. But I remember now - the difficulty is to get the class instance of this array type...

**Comment by jsotuyod** [comment]
@adangel nice! I didn't remember `Dimensionable` exists. Good call!

**Pull request merged**